src/runtime/sources.cpp

```diff
@@ -283,8 +284,13 @@ bool find_all_sources(Runtime &runtime, bool workspace) {
   int flags, dirfd = open(".", O_RDONLY);
   if ((flags = fcntl(dirfd, F_GETFD, 0)) != -1) fcntl(dirfd, F_SETFD, flags | FD_CLOEXEC);
 
-  std::vector<std::string> sources = scan_git(dirfd);
-  std::sort(sources.begin(), sources.end());
+  std::vector<std::string> sources;
+
+  // Source discovery requires git
+  if (is_readable(".git")) {
```

Comment:
Does `is_readable` work on folders?

Reply:
I'm pretty sure it does. It  was working for me when I was testing it.

Under the hood its just calling `access(path, R_OK)` https://man7.org/linux/man-pages/man2/access.2.html

Reply:
In that case, it might be missing a `X_OK` check as well?  I don't know exactly where the line between reading and executing a directory falls, but I know there's some muddiness between them.

Reply:
I think for true correctness, we actually do need to check `X_OK` on the directory. I'm think its okay to not do that here though since it would be very strange to have `R_OK` but not `X_OK` on the `.git` repo and a I'd have to make another function or just call `access` directly here since `is_readable` actively shouldn't check `X_OK`

Reply:
hmmm that said, directly calling `access` here isn't arduous. I should probably just do that

Reply:
I haven't read all of this, but something to be careful of is that sometimes the `.git` is a file rather than a directory

Reply:
I think we should just state the inode honestly. If the inode exists, we assume the user is using git

Reply:
That's actually a good point.  All we care about is whether git is tracking things or not -- if the workspace is set up to use git but git is broken, then it's not an issue on our side and git will be able to report the problem much better than we are.

Reply:
Done!